File: src/python/problems/AddTwoNumbers.py
# https://leetcode.com/problems/add-two-numbers/
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

class AddTwoNumbers:
    def Solution(self, l1: Optional[ListNode], l2: Optional[ListNode]) -> Optional[ListNode]:
        node1 = l1
        node2 = l2
        result = ListNode()
        current = result
        carry = 0

        while node1 or node2 or carry > 0:
            num1 = node1.val if node1 else 0
            num2 = node2.val if node2 else 0
            sum = carry + num1 + num2
            carry = sum // 10
            current.next = ListNode(sum % 10)
        
            current = current.next
            node1 = node1.next if node1 else None
            node2 = node2.next if node2 else None

        return result.next


AddTwoNumbers().Solution(ListNode(1, ListNode(2, ListNode(3))), ListNode(1, ListNode(2, ListNode(3))))

File: src/go/problems/AddTwoNumbers.go
// https://leetcode.com/problems/add-two-numbers/

package problems

func addTwoNumbers(l1 *ListNode, l2 *ListNode) *ListNode {
	node1 := l1
	node2 := l2
	result := &ListNode{}
	current := result
	var carry = 0

	for node1 != nil || node2 != nil || carry > 0 {
		var sum = carry
		if node1 != nil {
			sum = sum + node1.Val
			node1 = node1.Next
		}

		if node2 != nil {
			sum = sum + node2.Val
			node2 = node2.Next
		}

		carry = sum / 10
		current.Next = &ListNode{Val: sum % 10}
		current = current.Next
	}

	return result.Next
}
